feat: move spectral ruleset to source code

magicmatatjahu opened the pull request:
<!--   Thank you for your contribution. Before you submit the pull request:
1. Follow contributing guidelines and make sure PR title follows Conventional Commits specification -> https://github.com/asyncapi/generator/blob/master/CONTRIBUTING.md
2. Test your changes and attach their results to the pull request.
3. Update the relevant documentation.
-->

**Description**

- add Spectral ruleset to ParserJS source code.

**Related issue(s)**
Resolves https://github.com/asyncapi/parser-js/issues/698

jonaslagoni reviewed (commented):
Many of these ruleset functions are directly from Spectral right? or are they custom ones?

Arent there a lot of duplicate rulesets here that already exist in Spectral?

magicmatatjahu commented on src/ruleset/formats.ts:
```diff
@@ -0,0 +1,51 @@
+  isAas2(document) && aas2_5Regex.test(String((document as MaybeAsyncAPI).asyncapi));
+aas2_5.displayName = 'AsyncAPI 2.5.x';
+
+export const aas2_6: Format = (document: unknown): boolean =>
```
I know, but it will be release in two (max) days 😄

jonaslagoni replied:
Right, but just keep in mind that while it is likely that the version is introduced, you wouldn't be able to revert it later, because it's part of a larger commit.

So new features = new PR, at least in theory 😆

magicmatatjahu replied:
So I will wait for it :P Don't worry.

magicmatatjahu commented:
@jonaslagoni Yes these are ported rules from spectral itself (with this same code with slight changes), which by the way some I wrote myself as well as you :) I fixed some bugs that we observed, such as bad description of warnings (it was `must be...`, and I changed it to `should be...` etc), `operationId` is now optional (warning severity). In addition, I had to add rules that only work with the parser itself, e.g. related to custom schemas. I know that this may seem like a duplication of rules, but it is done consciously - I described the problem here https://github.com/asyncapi/parser-js/issues/698

I will add in the documentation that the rules were copied from Spectral's project and write why we decided to do so. WDYT? Is it enough, or you have a problem with that?

jonaslagoni commented:
@magicmatatjahu I find it a bit odd that its needed, these support functions are not something Spectral exports if they are required to write custom functions in TS 🤔?

My response in https://github.com/asyncapi/parser-js/issues/698 was not meant to be interpreted as all functions in spectral should be duplicated in the parser repo, but it's fine to add custom ones first here, and then contribute them back to spectral if other users need them.

It's entirely acceptable IMO that the rules live in spectral and we import them in the parser, in the worst case we can copy them to here if we need to rewrite them for whatever reason. Otherwise, the rules should live in spectral where possible 🙂 

Hope that clarifies my questions, otherwise let me know 🙂

magicmatatjahu commented:
@jonaslagoni And we misunderstood each other. In my issue (even looking at the title) I described that I would like to move the whole ruleset, not individual rules and have it integrated with our parser. Overriding/adding new rules is already possible in the parser itself and we do that, e.g. changing `asyncapi-operation-operationId` rule to the `warning` severity (and not with `error`).

jonaslagoni commented:
> Overriding/adding new rules is already possible in the parser itself and we do that, e.g. changing asyncapi-operation-operationId rule to the warning severity (and not with error).

And that's fine, if you need to change any of the existing spectral rules the change can start here in the parser repo.

But I don't see any benefit in moving all the rules if there is nothing in them that you wish to change 🤔 Do you see any benefit there?

Also what are your thoughts on the support functions?

magicmatatjahu commented:
I just want to avoid having sources in another repo and have everything in one. You will say that this is ok, but what in case someone contribute a rule we don't want and adds it with `error` severity, and it shouldn't be? It's better to have control of that in one repo. And if we have some rules at this repo and in the Spectral, wouldn't it be better to have it in the Spectral - here there are problems, which I described in the https://github.com/asyncapi/parser-js/issues/698

> Also what are your thoughts on the support functions?

What do you mean? Rules?

jonaslagoni commented:
> You will say that this is ok, but what in case someone contribute a rule we don't want and adds it with `error` severity, and it shouldn't be?

That's what you use the spectral configuration for, right? I.e. you disable them in the configuration (I am assuming you can easily do this in our setup)? 
```
extends: [spectral:asyncapi]
rules:
  asyncapi-new-rule-we-dont-want: off
```

> What do you mean? Rules?

The rulesets use a bunch of helper functions, those I mean 🙂

magicmatatjahu commented:
> That's what you use the spectral configuration for, right? I.e. you disable them in the configuration (I am assuming you can easily do this in our setup)?

I answered about that here:

> The other thing I wanted to get was to remove some of the rules, which at the moment I do not see the point of having (what Spectral ruleset has), and importing these rules from Spectral would be in the final bundle and I want to avoid that.

---

> The rulesets use a bunch of helper functions, those I mean 🙂

Yeah, we can reuse them, but they are only written to support one thing, e.g. checking uniqueness of `operationId`s, so I don't see any benefits to reuse them.

jonaslagoni commented:
@magicmatatjahu gonna continue in https://github.com/asyncapi/parser-js/issues/698

magicmatatjahu commented:
/rtm
